File: src/main/scala/com/kotori316/fluidtank/transport/PipeBlock.java
package com.kotori316.fluidtank.transport;

import java.util.Arrays;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.OptionalInt;
import java.util.function.Predicate;
import java.util.stream.Stream;

import com.google.common.collect.ImmutableBiMap;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.util.StringRepresentable;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.BlockItem;
import net.minecraft.world.item.DyeColor;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.LevelAccessor;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.EntityBlock;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.EnumProperty;
import net.minecraft.world.level.material.FluidState;
import net.minecraft.world.level.material.Material;
import net.minecraft.world.phys.AABB;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.Vec3;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.phys.shapes.Shapes;
import net.minecraft.world.phys.shapes.VoxelShape;
import org.apache.commons.lang3.tuple.Pair;
import org.jetbrains.annotations.NotNull;

import com.kotori316.fluidtank.Config;
import com.kotori316.fluidtank.FluidTank;
import com.kotori316.fluidtank.ModObjects;
import com.kotori316.fluidtank.Utils;

public abstract class PipeBlock extends Block implements EntityBlock {
    public static final VoxelShape BOX_AABB = Shapes.box(0.25, 0.25, 0.25, 0.75, 0.75, 0.75);
    public static final VoxelShape North_AABB = Shapes.box(0.25, 0.25, 0, 0.75, 0.75, 0.25);
    public static final VoxelShape South_AABB = Shapes.box(0.25, 0.25, .75, 0.75, 0.75, 1);
    public static final VoxelShape West_AABB = Shapes.box(0, 0.25, 0.25, .25, 0.75, 0.75);
    public static final VoxelShape East_AABB = Shapes.box(.75, 0.25, 0.25, 1, 0.75, 0.75);
    public static final VoxelShape UP_AABB = Shapes.box(0.25, .75, 0.25, 0.75, 1, 0.75);
    public static final VoxelShape Down_AABB = Shapes.box(0.25, 0, 0.25, 0.75, .25, 0.75);

    public static final EnumProperty<Connection> NORTH = EnumProperty.create("north", Connection.class);
    public static final EnumProperty<Connection> SOUTH = EnumProperty.create("south", Connection.class);
    public static final EnumProperty<Connection> WEST = EnumProperty.create("west", Connection.class);
    public static final EnumProperty<Connection> EAST = EnumProperty.create("east", Connection.class);
    public static final EnumProperty<Connection> UP = EnumProperty.create("up", Connection.class);
    public static final EnumProperty<Connection> DOWN = EnumProperty.create("down", Connection.class);
    private static final ImmutableBiMap<EnumProperty<Connection>, VoxelShape> SHAPE_MAP = ImmutableBiMap.of(
        PipeBlock.NORTH, North_AABB,
        PipeBlock.SOUTH, South_AABB,
        PipeBlock.WEST, West_AABB,
        PipeBlock.EAST, East_AABB,
        PipeBlock.UP, UP_AABB,
        PipeBlock.DOWN, Down_AABB
    );
    public static final ImmutableBiMap<Direction, EnumProperty<Connection>> FACING_TO_PROPERTY_MAP = ImmutableBiMap.of(
        Direction.NORTH, NORTH,
        Direction.SOUTH, SOUTH,
        Direction.WEST, WEST,
        Direction.EAST, EAST,
        Direction.UP, UP,
        Direction.DOWN, DOWN
    );

    @NotNull
    public static EnumProperty<Connection> getPropertyFromDirection(Direction facing) {
        return Objects.requireNonNull(FACING_TO_PROPERTY_MAP.get(facing));
    }

    public BlockItem itemBlock() {
        return blockItem;
    }

    private final BlockItem blockItem;
    public final ResourceLocation registryName;

    public PipeBlock() {
        super(Block.Properties.of(ModObjects.MATERIAL_PIPE())
            .strength(0.5f));
        registryName = new ResourceLocation(FluidTank.modID, getRegName());
        this.registerDefaultState(getStateDefinition().any()
                .setValue(NORTH, Connection.NO_CONNECTION)
                .setValue(SOUTH, Connection.NO_CONNECTION)
                .setValue(WEST, Connection.NO_CONNECTION)
                .setValue(EAST, Connection.NO_CONNECTION)
                .setValue(UP, Connection.NO_CONNECTION)
                .setValue(DOWN, Connection.NO_CONNECTION)
//            .with(WATERLOGGED, false)
        );
        blockItem = new BlockItem(this, new Item.Properties());
    }

    @NotNull
    protected abstract String getRegName();

    protected abstract boolean isHandler(BlockGetter level, BlockPos pos, EnumProperty<Connection> property);

    @Override
    public abstract BlockEntity newBlockEntity(BlockPos pos, BlockState state);

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        builder.add(NORTH, EAST, SOUTH, WEST, UP, DOWN/*, WATERLOGGED*/);
    }

    @Override
    @SuppressWarnings("deprecation")
    public VoxelShape getShape(BlockState state, BlockGetter worldIn, BlockPos pos, CollisionContext context) {
        return SHAPE_MAP.entrySet().stream()
            .filter(s -> state.getValue(s.getKey()) != Connection.NO_CONNECTION || isHandler(worldIn, pos, s.getKey()))
            .map(Map.Entry::getValue)
            .reduce(BOX_AABB, Shapes::or);
    }

    @Override
    public BlockState getStateForPlacement(BlockPlaceContext context) {
        Level worldIn = context.getLevel();
        BlockPos pos = context.getClickedPos();
        return this.defaultBlockState()
            .setValue(NORTH, canConnectTo(worldIn, pos.north(), Direction.NORTH))
            .setValue(EAST, canConnectTo(worldIn, pos.east(), Direction.EAST))
            .setValue(SOUTH, canConnectTo(worldIn, pos.south(), Direction.SOUTH))
            .setValue(WEST, canConnectTo(worldIn, pos.west(), Direction.WEST))
            .setValue(DOWN, canConnectTo(worldIn, pos.below(), Direction.DOWN))
            .setValue(UP, canConnectTo(worldIn, pos.above(), Direction.UP));
//            .with(WATERLOGGED, fluidState.getFluid() == Fluids.WATER);
    }

    @Override
    @SuppressWarnings("deprecation")
    public BlockState updateShape(BlockState stateIn, Direction facing, BlockState facingState,
                                  LevelAccessor worldIn, BlockPos currentPos, BlockPos facingPos) {
        /*if (stateIn.get(WATERLOGGED)) {
            worldIn.getPendingFluidTicks().scheduleTick(currentPos, Fluids.WATER, Fluids.WATER.getTickRate(worldIn));
        }*/
        Connection value;
        Connection now = stateIn.getValue(getPropertyFromDirection(facing));
        if (facingState.getBlock() == this) {
            value = facingState.getValue(getPropertyFromDirection(facing.getOpposite()));
            return stateIn.setValue(getPropertyFromDirection(facing), value);
        } else {
            value = canConnectTo(worldIn, currentPos.relative(facing), facing);
            if (value.is(Connection.NO_CONNECTION)) {
                if (facingState.getMaterial() == Material.AIR || facingState.getMaterial().isLiquid()) {
                    return stateIn.setValue(getPropertyFromDirection(facing), value);
                } else {
                    return stateIn;
                }
            } else if (value.hasConnection() ^ now.hasConnection())
                return stateIn.setValue(getPropertyFromDirection(facing), value);
            else
                return stateIn;
        }
    }

    private Connection canConnectTo(BlockGetter level, BlockPos pos, Direction direction) {
        BlockState blockState = level.getBlockState(pos);
        BlockEntity entity = level.getBlockEntity(pos);
        if (blockState.getBlock() == this) {
            if (!Config.content().enablePipeRainbowRenderer().get() && entity instanceof PipeTileBase p) {
                return p.getColor() == Config.content().pipeColor().get() ? Connection.CONNECTED : Connection.NO_CONNECTION;
            }
            return Connection.CONNECTED;
        } else {
            if (entity != null) {
                return getConnection(direction, entity);
            } else {
                return Connection.NO_CONNECTION;
            }
        }
    }

    @NotNull
    protected abstract Connection getConnection(Direction direction, @NotNull BlockEntity entity);

    @Override
    @SuppressWarnings("deprecation")
    public InteractionResult use(BlockState state, Level worldIn, BlockPos pos, Player player, InteractionHand handIn, BlockHitResult hit) {
        if (player.getItemInHand(handIn).getItem() instanceof BlockItem || player.isCrouching())
            return InteractionResult.PASS;
        // Dying pipe.
        OptionalInt maybeColor = Utils.getItemColor(player.getItemInHand(handIn));
        if (maybeColor.isPresent() && !Config.content().enablePipeRainbowRenderer().get()) {
            if (!worldIn.isClientSide) {
                Optional.ofNullable(worldIn.getBlockEntity(pos)).map(PipeTileBase.class::cast).ifPresent(p -> p.changeColor(maybeColor.getAsInt()));
                Object colorName = Stream.of(DyeColor.values()).filter(d -> d.getMaterialColor().col == maybeColor.getAsInt()).findFirst()
                    .map(c -> (Object) Component.translatable("color.minecraft." + c))
                    .orElse(String.format("#%06x", maybeColor.getAsInt()));
                player.displayClientMessage(
                    Component.translatable("chat.fluidtank.change_color", colorName),
                    false);
            }
            return InteractionResult.SUCCESS;
        }

        // Modifying pipe connection.
        Vec3 d = hit.getLocation().subtract(pos.getX(), pos.getY(), pos.getZ());
        Predicate<Map.Entry<?, VoxelShape>> predicate = e -> {
            AABB box = e.getValue().bounds();
            return box.minX <= d.x && box.maxX >= d.x && box.minY <= d.y && box.maxY >= d.y && box.minZ <= d.z && box.maxZ >= d.z;
        };
        Optional<Pair<BlockState, Boolean>> blockState = SHAPE_MAP.entrySet().stream()
            .filter(predicate)
            .map(Map.Entry::getKey)
            .findFirst()
            .map(p -> {
                if (worldIn.getBlockState(pos.relative(Objects.requireNonNull(FACING_TO_PROPERTY_MAP.inverse().get(p)))).getBlock() != this)
                    return Pair.of(state.cycle(p), false);
                else
                    return Pair.of(state.setValue(p, Connection.onOffConnection(state.getValue(p))), true);
            });
        if (blockState.isPresent()) {
            if (!worldIn.isClientSide) {
                worldIn.setBlockAndUpdate(pos, blockState.get().getKey());
                if (blockState.get().getValue())
                    Optional.ofNullable(worldIn.getBlockEntity(pos)).map(PipeTileBase.class::cast).ifPresent(PipeTileBase::connectorUpdate);
            }
            return InteractionResult.SUCCESS;
        } else {
            return super.use(state, worldIn, pos, player, handIn, hit);
        }
    }

    @Override
    @SuppressWarnings("deprecation")
    public void neighborChanged(BlockState state, Level worldIn, BlockPos pos, Block blockIn, BlockPos fromPos, boolean isMoving) {
        super.neighborChanged(state, worldIn, pos, blockIn, fromPos, isMoving);
        BlockState fromState = worldIn.getBlockState(fromPos);
        // Update connection between pipes.
        if (fromState.getBlock() == this) {
            BlockPos vec = fromPos.subtract(pos);
            Direction direction = Direction.fromNormal(vec.getX(), vec.getY(), vec.getZ());
            if (direction != null) {
                if (fromState.getValue(getPropertyFromDirection(direction.getOpposite())) == Connection.NO_CONNECTION) {
                    worldIn.setBlockAndUpdate(pos, state.setValue(getPropertyFromDirection(direction), Connection.NO_CONNECTION));
                } else if (fromState.getValue(getPropertyFromDirection(direction.getOpposite())) == Connection.CONNECTED) {
                    worldIn.setBlockAndUpdate(pos, state.setValue(getPropertyFromDirection(direction), Connection.CONNECTED));
                }
            }
        }
        // Update handlers
        if (!worldIn.isClientSide) {
            Optional.ofNullable((PipeTileBase) worldIn.getBlockEntity(pos))
                .ifPresent(t -> t.removeCapCache(fromPos));
        }
    }

    @Override
    @SuppressWarnings("deprecation")
    public void onRemove(BlockState state, Level level, BlockPos pos, BlockState newState, boolean moved) {
        if (state.getBlock() != newState.getBlock()) {
            BlockEntity entity = level.getBlockEntity(pos);
            if (entity instanceof PipeTileBase tile) {
                tile.connectorUpdate();
            }
            super.onRemove(state, level, pos, newState, moved);
        }
    }

    @Override
    @SuppressWarnings("deprecation")
    public FluidState getFluidState(BlockState state) {
        return /*state.get(WATERLOGGED) ? Fluids.WATER.getStillFluidState(false) :*/ super.getFluidState(state);
    }

    public enum Connection implements StringRepresentable {
        NO_CONNECTION,
        CONNECTED,
        INPUT,
        OUTPUT;

        public String getName() {
            return name().toLowerCase();
        }

        public boolean is(Connection c1, Connection... cs) {
            return this == c1 || Arrays.asList(cs).contains(this);
        }

        public boolean hasConnection() {
            return is(CONNECTED, INPUT, OUTPUT);
        }

        public boolean isOutput() {
            return is(OUTPUT);
        }

        public boolean isInput() {
            return is(INPUT);
        }

        public static Connection onOffConnection(Connection now) {
            if (now == NO_CONNECTION)
                return CONNECTED;
            else
                return NO_CONNECTION;
        }

        @Override
        public String getSerializedName() {
            return getName();
        }
    }

}

File: src/main/scala/com/kotori316/fluidtank/transport/PipeBlock.scala
package com.kotori316.fluidtank.transport

import com.google.common.collect.ImmutableBiMap
import com.kotori316.fluidtank.{BlockPosHelper, Config, FluidTank, ModObjects, Utils}
import net.minecraft.core.{BlockPos, Direction}
import net.minecraft.network.chat.{Component, TextComponent, TranslatableComponent}
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.context.BlockPlaceContext
import net.minecraft.world.item.{BlockItem, DyeColor, Item}
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.state.properties.EnumProperty
import net.minecraft.world.level.block.state.{BlockBehaviour, BlockState, StateDefinition}
import net.minecraft.world.level.block.{Block, EntityBlock}
import net.minecraft.world.level.material.{FluidState, Material}
import net.minecraft.world.level.{BlockGetter, Level, LevelAccessor}
import net.minecraft.world.phys.BlockHitResult
import net.minecraft.world.phys.shapes.{CollisionContext, Shapes, VoxelShape}
import net.minecraft.world.{InteractionHand, InteractionResult}

import scala.annotation.nowarn
import scala.jdk.CollectionConverters._

object PipeBlock {
  val BOX_AABB: VoxelShape = Shapes.box(0.25, 0.25, 0.25, 0.75, 0.75, 0.75)
  val North_AABB: VoxelShape = Shapes.box(0.25, 0.25, 0, 0.75, 0.75, 0.25)
  val South_AABB: VoxelShape = Shapes.box(0.25, 0.25, .75, 0.75, 0.75, 1)
  val West_AABB: VoxelShape = Shapes.box(0, 0.25, 0.25, .25, 0.75, 0.75)
  val East_AABB: VoxelShape = Shapes.box(.75, 0.25, 0.25, 1, 0.75, 0.75)
  val UP_AABB: VoxelShape = Shapes.box(0.25, .75, 0.25, 0.75, 1, 0.75)
  val Down_AABB: VoxelShape = Shapes.box(0.25, 0, 0.25, 0.75, .25, 0.75)
  val NORTH: EnumProperty[PipeBlockConnection] = EnumProperty.create("north", classOf[PipeBlockConnection])
  val SOUTH: EnumProperty[PipeBlockConnection] = EnumProperty.create("south", classOf[PipeBlockConnection])
  val WEST: EnumProperty[PipeBlockConnection] = EnumProperty.create("west", classOf[PipeBlockConnection])
  val EAST: EnumProperty[PipeBlockConnection] = EnumProperty.create("east", classOf[PipeBlockConnection])
  val UP: EnumProperty[PipeBlockConnection] = EnumProperty.create("up", classOf[PipeBlockConnection])
  val DOWN: EnumProperty[PipeBlockConnection] = EnumProperty.create("down", classOf[PipeBlockConnection])
  private val SHAPE_MAP = ImmutableBiMap.of(
    PipeBlock.NORTH, North_AABB,
    PipeBlock.SOUTH, South_AABB,
    PipeBlock.WEST, West_AABB,
    PipeBlock.EAST, East_AABB,
    PipeBlock.UP, UP_AABB,
    PipeBlock.DOWN, Down_AABB,
  )
  val FACING_TO_PROPERTY_MAP: ImmutableBiMap[Direction, EnumProperty[PipeBlockConnection]] = ImmutableBiMap.of(
    Direction.NORTH, NORTH,
    Direction.SOUTH, SOUTH,
    Direction.WEST, WEST,
    Direction.EAST, EAST,
    Direction.UP, UP,
    Direction.DOWN, DOWN,
  )

  def getPropertyFromDirection(facing: Direction): EnumProperty[PipeBlockConnection] = FACING_TO_PROPERTY_MAP.get(facing)
}

abstract class PipeBlock extends Block(BlockBehaviour.Properties.of(ModObjects.MATERIAL_PIPE).strength(0.5f)) with EntityBlock {
  setRegistryName(FluidTank.modID, getRegName)
  this.registerDefaultState(getStateDefinition.any
    .setValue(PipeBlock.NORTH, PipeBlockConnection.NO_CONNECTION)
    .setValue(PipeBlock.SOUTH, PipeBlockConnection.NO_CONNECTION)
    .setValue(PipeBlock.WEST, PipeBlockConnection.NO_CONNECTION)
    .setValue(PipeBlock.EAST, PipeBlockConnection.NO_CONNECTION)
    .setValue(PipeBlock.UP, PipeBlockConnection.NO_CONNECTION)
    .setValue(PipeBlock.DOWN, PipeBlockConnection.NO_CONNECTION)
    //            .with(WATERLOGGED, false)
  )
  private final val blockItem = new BlockItem(this, new Item.Properties().tab(ModObjects.CREATIVE_TABS))
  blockItem.setRegistryName(FluidTank.modID, getRegName)

  def itemBlock: BlockItem = blockItem

  protected def getRegName: String

  protected def isHandler(level: BlockGetter, pos: BlockPos, property: EnumProperty[PipeBlockConnection]): Boolean

  override def newBlockEntity(pos: BlockPos, state: BlockState): BlockEntity

  override protected def createBlockStateDefinition(builder: StateDefinition.Builder[Block, BlockState]): Unit = {
    builder.add(PipeBlock.NORTH, PipeBlock.EAST, PipeBlock.SOUTH, PipeBlock.WEST, PipeBlock.UP, PipeBlock.DOWN /*, WATERLOGGED*/)
  }

  //noinspection ScalaDeprecation
  override def getShape(state: BlockState, worldIn: BlockGetter, pos: BlockPos, context: CollisionContext): VoxelShape = {
    PipeBlock.SHAPE_MAP.entrySet.asScala
      .filter(e => state.getValue(e.getKey) != PipeBlockConnection.NO_CONNECTION || isHandler(worldIn, pos, e.getKey))
      .map(_.getValue)
      .fold(PipeBlock.BOX_AABB)(Shapes.or)
  }

  override def getStateForPlacement(context: BlockPlaceContext): BlockState = {
    val worldIn = context.getLevel
    val pos = context.getClickedPos
    this.defaultBlockState
      .setValue(PipeBlock.NORTH, canConnectTo(worldIn, pos.north, Direction.NORTH))
      .setValue(PipeBlock.EAST, canConnectTo(worldIn, pos.east, Direction.EAST))
      .setValue(PipeBlock.SOUTH, canConnectTo(worldIn, pos.south, Direction.SOUTH))
      .setValue(PipeBlock.WEST, canConnectTo(worldIn, pos.west, Direction.WEST))
      .setValue(PipeBlock.DOWN, canConnectTo(worldIn, pos.below, Direction.DOWN))
      .setValue(PipeBlock.UP, canConnectTo(worldIn, pos.above, Direction.UP))
    //            .with(WATERLOGGED, fluidState.getFluid() == Fluids.WATER);
  }

  //noinspection ScalaDeprecation
  override def updateShape(stateIn: BlockState, facing: Direction, facingState: BlockState, worldIn: LevelAccessor, currentPos: BlockPos, facingPos: BlockPos): BlockState = {
    /*if (stateIn.get(WATERLOGGED)) {
               worldIn.getPendingFluidTicks().scheduleTick(currentPos, Fluids.WATER, Fluids.WATER.getTickRate(worldIn));
     }*/
    val now = stateIn.getValue(PipeBlock.getPropertyFromDirection(facing))
    if (facingState.getBlock == this) {
      val value = facingState.getValue(PipeBlock.getPropertyFromDirection(facing.getOpposite))
      stateIn.setValue(PipeBlock.getPropertyFromDirection(facing), value)
    } else {
      val value = canConnectTo(worldIn, currentPos.relative(facing), facing)
      if (value is PipeBlockConnection.NO_CONNECTION) {
        if (facingState.getMaterial == Material.AIR || facingState.getMaterial.isLiquid) {
          stateIn.setValue(PipeBlock.getPropertyFromDirection(facing), value)
        } else {
          stateIn
        }
      } else if (value.hasConnection ^ now.hasConnection) {
        stateIn.setValue(PipeBlock.getPropertyFromDirection(facing), value)
      } else {
        stateIn
      }
    }
  }

  private def canConnectTo(level: BlockGetter, pos: BlockPos, direction: Direction): PipeBlockConnection = {
    val blockState = level.getBlockState(pos)
    val entity = level.getBlockEntity(pos)
    if (blockState.getBlock == this) {
      entity match {
        case p: PipeTileBase if !Config.content.enablePipeRainbowRenderer.get() =>
          if (p.getColor == Config.content.pipeColor.get) PipeBlockConnection.CONNECTED
          else PipeBlockConnection.NO_CONNECTION
        case _ => PipeBlockConnection.CONNECTED
      }
    } else {
      if (entity != null) getConnection(direction, entity)
      else PipeBlockConnection.NO_CONNECTION
    }
  }

  protected def getConnection(direction: Direction, entity: BlockEntity): PipeBlockConnection

  private def getPipeTile(worldIn: Level, pos: BlockPos): Option[PipeTileBase] =
    Option(worldIn.getBlockEntity(pos)).collect { case t: PipeTileBase => t }

  //noinspection ScalaDeprecation
  override def use(state: BlockState, worldIn: Level, pos: BlockPos, player: Player, handIn: InteractionHand, hit: BlockHitResult): InteractionResult = {
    if (player.getItemInHand(handIn).getItem.isInstanceOf[BlockItem] || player.isCrouching)
      return InteractionResult.PASS
    // Dying pipe.
    val maybeColor = Utils.getItemColor(player.getItemInHand(handIn))
    if (maybeColor.isPresent && !Config.content.enablePipeRainbowRenderer.get()) {
      if (!worldIn.isClientSide) {
        getPipeTile(worldIn, pos).foreach(_.changeColor(maybeColor.getAsInt))
        val colorName: Component = DyeColor.values.find(_.getMaterialColor.col == maybeColor.getAsInt).map(c => new TranslatableComponent("color.minecraft." + c))
          .getOrElse(new TextComponent(String.format("#%06x", maybeColor.getAsInt)))
        player.displayClientMessage(new TranslatableComponent("chat.fluidtank.change_color", colorName), false)
      }
      return InteractionResult.SUCCESS
    }

    // Modifying pipe connection.
    val d = hit.getLocation.subtract(pos.getX, pos.getY, pos.getZ)
    val blockState = PipeBlock.SHAPE_MAP.entrySet.asScala
      .filter { e =>
        val box = e.getValue.bounds
        box.minX <= d.x && box.maxX >= d.x && box.minY <= d.y && box.maxY >= d.y && box.minZ <= d.z && box.maxZ >= d.z
      }
      .map(_.getKey)
      .headOption
      .map { p =>
        if (worldIn.getBlockState(pos.relative(PipeBlock.FACING_TO_PROPERTY_MAP.inverse.get(p))).getBlock != this)
          (state.cycle(p), false)
        else (state.setValue(p, PipeBlockConnection.onOffConnection(state.getValue(p))), true)
      }
    blockState match {
      case Some((newState, update)) =>
        if (!worldIn.isClientSide) {
          worldIn.setBlockAndUpdate(pos, newState)
          if (update)
            getPipeTile(worldIn, pos).foreach(_.connectorUpdate())
        }
        InteractionResult.SUCCESS
      case None =>
        super.use(state, worldIn, pos, player, handIn, hit): @nowarn("cat=deprecation")
    }
  }

  //noinspection ScalaDeprecation
  override def neighborChanged(state: BlockState, worldIn: Level, pos: BlockPos, blockIn: Block, fromPos: BlockPos, isMoving: Boolean): Unit = {
    super.neighborChanged(state, worldIn, pos, blockIn, fromPos, isMoving): @nowarn("cat=deprecation")
    val fromState = worldIn.getBlockState(fromPos)
    // Update connection between pipes.
    if (fromState.getBlock != this) {
      val vec = fromPos.sub(pos)
      val direction = Direction.fromNormal(vec.getX, vec.getY, vec.getZ)
      if (direction != null) {
        if (fromState.getValue(PipeBlock.getPropertyFromDirection(direction.getOpposite)) == PipeBlockConnection.NO_CONNECTION)
          worldIn.setBlockAndUpdate(pos, state.setValue(PipeBlock.getPropertyFromDirection(direction), PipeBlockConnection.NO_CONNECTION))
        else if (fromState.getValue(PipeBlock.getPropertyFromDirection(direction.getOpposite)) == PipeBlockConnection.CONNECTED)
          worldIn.setBlockAndUpdate(pos, state.setValue(PipeBlock.getPropertyFromDirection(direction), PipeBlockConnection.CONNECTED))
      }
    }
    // Update handlers
    if (!worldIn.isClientSide) {
      getPipeTile(worldIn, pos).foreach(_.removeCapCache(fromPos))
    }
  }

  //noinspection ScalaDeprecation
  override def onRemove(state: BlockState, level: Level, pos: BlockPos, newState: BlockState, moved: Boolean): Unit =
    if (state.getBlock != newState.getBlock) {
      getPipeTile(level, pos).foreach(_.connectorUpdate())
      super.onRemove(state, level, pos, newState, moved): @nowarn("cat=deprecation")
    }

  //noinspection ScalaDeprecation
  override def getFluidState(state: BlockState): FluidState = {
    /*state.get(WATERLOGGED) ? Fluids.WATER.getStillFluidState(false) :*/
    super.getFluidState(state): @nowarn("cat=deprecation")
  }
}